cmd/zrepl/zrepl.c

```diff
@@ -0,0 +1,976 @@
+		ZREPL_LOG("Connection to TGT controller iss successful\n");
+	}
+
+	efd = epoll_create1(0);
+	if (efd == -1) {
+		ZREPL_ERRLOG("epoll_create() failed with errno:%d\n", errno);
+		goto exit;
+	}
+
+	event.data.fd = sfd;
+	event.events = EPOLLIN | EPOLLET | EPOLLERR |
+	    EPOLLHUP | EPOLLRDHUP;
+	rc = epoll_ctl(efd, EPOLL_CTL_ADD, sfd, &event);
+	if (rc == -1) {
+		ZREPL_ERRLOG("epoll_ctl() failed with errno:%d\n", errno);
+		goto exit;
+	}
+
+	/* Buffer where events are returned */
+	events = calloc(MAXEVENTS, sizeof (event));
+
+	/* The event loop */
+	while (1) {
+		int n, i;
+		n = epoll_wait(efd, events, MAXEVENTS, -1);
+		for (i = 0; i < n; i++) {
+			if ((events[i].events & EPOLLERR) ||
+			    (events[i].events & EPOLLHUP) ||
+			    (events[i].events & EPOLLRDHUP)) {
+				/*
+				 * Error has occurred on this socket
+				 * close it and open a new socket after
+				 * 5 sec of sleep.
+				 */
+				ZREPL_ERRLOG("epoll err() :%d\n", errno);
+				close(events[i].data.fd);
+				printf("Retrying ....\n");
+				sleep(5);
+				sfd = create_and_bind(mgmt_port);
+				if (sfd == -1) {
+					goto exit;
+				}
+
+				rc = make_socket_non_blocking(sfd);
+				if (rc == -1) {
+					goto exit;
+				}
+retry1:
+				rc = connect(sfd,
+				    (struct sockaddr *)&istgt_addr,
+				    sizeof (istgt_addr));
+				if ((rc == -1) && (errno == EINTR)) {
+					ZREPL_ERRLOG("Failed to connect to"
+					    " istgt_controller with err:%d\n",
+					    errno);
+					sleep(2);
+					goto retry1;
+				}
+
+				event.data.fd = sfd;
+				event.events = EPOLLIN | EPOLLET |
+				    EPOLLERR | EPOLLHUP | EPOLLRDHUP;
+				rc = epoll_ctl(efd, EPOLL_CTL_ADD, sfd, &event);
+				if (rc == -1) {
+					ZREPL_ERRLOG("epoll_ctl() failed with"
+					    " errno:%d\n", errno);
+					goto exit;
+				}
+				continue;
+			}
+
+			bzero(&hdr, sizeof (hdr));
+			count = read(events[i].data.fd, (char *)&hdr,
+			    sizeof (hdr));
+			if (count == -1) {
+				ZREPL_ERRLOG("Read from socket failed"
+				    " with err:%d\n", errno);
+				goto exit;
+			}
+
+			if (hdr.opcode == ZVOL_OPCODE_HANDSHAKE) {
+				name = kmem_alloc(
+				    hdr.len * sizeof (char), KM_SLEEP);
+				count = read(events[i].data.fd, name,
+				    sizeof (char) * hdr.len);
+				if (count == -1) {
+					ZREPL_ERRLOG("Read from socket failed"
+					    " with err:%d\n", errno);
+					goto exit;
+				}
+
+				rc = uzfs_zvol_mgmt_do_handshake(&hdr,
+				    events[i].data.fd, name);
+				free(name);
+				if (rc == -1) {
+					ZREPL_ERRLOG("handshake failed with"
+					    " errno:%d\n", errno);
+					goto exit;
+				}
+			}
+		}
+	}
+exit:
+	if (events != NULL) {
+		free(events);
+	}
+	printf("uzfs_zvol_mgmt_thread thread exiting\n");
+	zk_thread_exit();
+}
+
+/*
+ * One thread per replica. Responsible for accepting
+ * IO connections. This thread will accept a connection
+ * and spawn a new thread for each new connection req.
+ */
+static void
+uzfs_zvol_io_conn_acceptor(void)
```

Comment:
I see a problem when this thread terminates for whatever reason it can be. the main thread does not know about it and the program will continue to run but will not accept any new connections. I think it would have been better if acceptor was run on behalf on the main thread instead of dedicating a separate thread to it, which makes the synchronization more difficult.

Reply:
Jan, this is by design, we know that if uzfs_zvol_io_conn_acceptor thread die down because of any reason, replica will not be able to accept new IO connections. Same goes with uzfs_zvol_mgmt_thread. I would discuss it with team here and get back to you. I am keeping this as it is for now until we have better way of doing things. 

One simple approach would be to kill main replica process if any of uzfs_zvol_mgmt_thread or uzfs_zvol_io_conn_acceptor goes down. That way container will reboot system back.

Reply:
👍